File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/administrativ/X001Mapper.java
package no.nav.melosys.eessi.service.sed.mapper.til_sed.administrativ;

import java.time.LocalDate;

import no.nav.melosys.eessi.models.SedType;
import no.nav.melosys.eessi.models.sed.Konstanter;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.nav.*;

import static no.nav.melosys.eessi.models.sed.Konstanter.*;


public class X001Mapper implements AdministrativSedMapper {

    public SED mapFraSed(SED sed, String aarsak, Boolean erCDM4_3) {
        SED x001 = new SED();
        x001.setSedType(SedType.X001.toString());
        x001.setSedGVer(DEFAULT_SED_G_VER);
        x001.setSedVer(erCDM4_3 ? SED_VER_CDM_4_3 : DEFAULT_SED_VER);
        x001.setNav(mapNav(sed, aarsak));

        return x001;
    }

    private Nav mapNav(SED sed, String aarsak) {
        Nav nav = new Nav();
        nav.setSak(mapSak(sed, aarsak));
        return nav;
    }

    private Sak mapSak(SED sed, String aarsak) {
        Sak sak = new Sak();
        sak.setAnmodning(mapAnmodning(aarsak));
        sak.setKontekst(mapKontekst(sed));
        return sak;
    }

    private Kontekst mapKontekst(SED sed) {
        Kontekst kontekst = new Kontekst();
        kontekst.setBruker(sed.getNav().getBruker());
        return kontekst;
    }

    private X001Anmodning mapAnmodning(String aarsakType) {
        X001Anmodning anmodning = new X001Anmodning();

        Aarsak aarsak = new Aarsak();
        aarsak.setType(aarsakType);

        Avslutning avslutning = new Avslutning();
        avslutning.setDato(LocalDate.now().format(Konstanter.dateTimeFormatter));
        avslutning.setAarsak(aarsak);
        avslutning.setType("automatisk");

        anmodning.setAvslutning(avslutning);
        return anmodning;
    }

    @Override
    public SedType getSedType() {
        return SedType.X001;
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/administrativ/X001Mapper.kt
package no.nav.melosys.eessi.service.sed.mapper.til_sed.administrativ

import io.getunleash.Unleash
import no.nav.melosys.eessi.config.featuretoggle.ToggleName.CDM_4_4
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.exception.MappingException
import no.nav.melosys.eessi.models.sed.Konstanter
import no.nav.melosys.eessi.models.sed.Konstanter.SED_VER_CDM_4_3
import no.nav.melosys.eessi.models.sed.Konstanter.SED_VER_CDM_4_4
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.nav.*
import java.time.LocalDate

class X001Mapper(private val unleash: Unleash) : AdministrativSedMapper {

    override fun getSedType(): SedType = SedType.X001

    fun mapFraSed(sed: SED, aarsak: String) = SED(
        sedType = SedType.X001.toString(),
        sedGVer = Konstanter.DEFAULT_SED_G_VER,
        sedVer = if (unleash.isEnabled(CDM_4_4)) SED_VER_CDM_4_4 else SED_VER_CDM_4_3,
        nav = mapNav(sed, aarsak)
    )

    private fun mapNav(sed: SED, aarsak: String) = Nav(
        sak = mapSak(sed, aarsak)
    )

    private fun mapSak(sed: SED, aarsak: String): Sak = Sak(
        anmodning = mapAnmodning(aarsak),
        kontekst = mapKontekst(sed)
    )

    private fun mapKontekst(sed: SED): Kontekst = Kontekst(
        bruker = sed.nav?.bruker ?: throw MappingException("nav.bruker er påkrevd for X001"),
    )

    private fun mapAnmodning(aarsakType: String) = X001Anmodning(
        avslutning = Avslutning(
            dato = LocalDate.now().format(Konstanter.dateTimeFormatter),
            aarsak = Aarsak(type = aarsakType),
            type = "automatisk"
        )
    )
}
